// @nx/react/src/generators/redux/redux.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.reduxGenerator = reduxGenerator;
exports.reduxGeneratorInternal = reduxGeneratorInternal;
const path = require("path");
const ast_utils_1 = require("../../utils/ast-utils");
const versions_1 = require("../../utils/versions");
const devkit_1 = require("@nx/devkit");
const js_1 = require("@nx/js");
const ensure_typescript_1 = require("@nx/js/src/utils/typescript/ensure-typescript");
const artifact_name_and_directory_utils_1 = require("@nx/devkit/src/generators/artifact-name-and-directory-utils");
let tsModule;
async function reduxGenerator(host, schema) {
    return reduxGeneratorInternal(host, {
        nameAndDirectoryFormat: 'derived',
        ...schema,
    });
}
async function reduxGeneratorInternal(host, schema) {
    const options = await normalizeOptions(host, schema);
    generateReduxFiles(host, options);
    addExportsToBarrel(host, options);
    const installTask = addReduxPackageDependencies(host);
    addStoreConfiguration(host, options);
    updateReducerConfiguration(host, options);
    await (0, devkit_1.formatFiles)(host);
    return installTask;
}
function generateReduxFiles(host, options) {
    (0, devkit_1.generateFiles)(host, (0, devkit_1.joinPathFragments)(__dirname, './files'), options.directory, {
        ...options,
        tmpl: '',
    });
    if (options.js) {
        (0, devkit_1.toJS)(host);
    }
}
function addReduxPackageDependencies(host) {
    return (0, devkit_1.addDependenciesToPackageJson)(host, {
        '@reduxjs/toolkit': versions_1.reduxjsToolkitVersion,
        'react-redux': versions_1.reactReduxVersion,
    }, {});
}
function addExportsToBarrel(host, options) {
    if (!tsModule) {
        tsModule = (0, ensure_typescript_1.ensureTypescript)();
    }
    const indexFilePath = path.join(options.projectSourcePath, options.js ? 'index.js' : 'index.ts');
    const indexSource = host.read(indexFilePath, 'utf-8');
    if (indexSource !== null) {
        const indexSourceFile = tsModule.createSourceFile(indexFilePath, indexSource, tsModule.ScriptTarget.Latest, true);
        const statePath = options.directory
            ? `./lib/${options.directory}/${options.fileName}`
            : `./lib/${options.fileName}`;
        const changes = (0, devkit_1.applyChangesToString)(indexSource, (0, ast_utils_1.addImport)(indexSourceFile, `export * from '${statePath}.slice';`));
        host.write(indexFilePath, changes);
    }
}
function addStoreConfiguration(host, options) {
    if (!options.appProjectSourcePath) {
        return;
    }
    const mainSource = host.read(options.appMainFilePath, 'utf-8');
    if (!mainSource.includes('redux')) {
        const mainSourceFile = tsModule.createSourceFile(options.appMainFilePath, mainSource, tsModule.ScriptTarget.Latest, true);
        const changes = (0, devkit_1.applyChangesToString)(mainSource, (0, ast_utils_1.addReduxStoreToMain)(options.appMainFilePath, mainSourceFile));
        host.write(options.appMainFilePath, changes);
    }
}
function updateReducerConfiguration(host, options) {
    if (!tsModule) {
        tsModule = (0, ensure_typescript_1.ensureTypescript)();
    }
    if (!options.appProjectSourcePath) {
        return;
    }
    const mainSource = host.read(options.appMainFilePath, 'utf-8');
    const mainSourceFile = tsModule.createSourceFile(options.appMainFilePath, mainSource, tsModule.ScriptTarget.Latest, true);
    const changes = (0, devkit_1.applyChangesToString)(mainSource, (0, ast_utils_1.updateReduxStore)(options.appMainFilePath, mainSourceFile, {
        keyName: `${options.constantName}_FEATURE_KEY`,
        reducerName: `${options.propertyName}Reducer`,
        modulePath: `${options.projectModulePath}`,
    }));
    host.write(options.appMainFilePath, changes);
}
async function normalizeOptions(host, options) {
    const { artifactName: name, directory, fileName, project: projectName, } = await (0, artifact_name_and_directory_utils_1.determineArtifactNameAndDirectoryOptions)(host, {
        artifactType: 'slice',
        callingGenerator: '@nx/react:redux',
        name: options.name,
        directory: options.directory,
        derivedDirectory: options.directory,
        flat: true,
        nameAndDirectoryFormat: options.nameAndDirectoryFormat,
        project: options.project,
        fileExtension: 'tsx',
    });
    let appProjectSourcePath;
    let appMainFilePath;
    const extraNames = (0, devkit_1.names)(name);
    const projects = (0, devkit_1.getProjects)(host);
    const project = projects.get(projectName);
    const { sourceRoot, projectType } = project;
    const tsConfigJson = (0, devkit_1.readJson)(host, (0, js_1.getRootTsConfigPathInTree)(host));
    const tsPaths = tsConfigJson.compilerOptions
        ? tsConfigJson.compilerOptions.paths || {}
        : {};
    const modulePath = projectType === 'application'
        ? options.directory
            ? `./app/${options.directory}/${extraNames.fileName}.slice`
            : `./app/${extraNames.fileName}.slice`
        : Object.keys(tsPaths).find((k) => tsPaths[k].some((s) => s.includes(sourceRoot)));
    // If --project is set to an app, automatically configure store
    // for it without needing to specify --appProject.
    options.appProject =
        options.appProject ||
            (projectType === 'application' ? projectName : undefined);
    if (options.appProject) {
        const appConfig = projects.get(options.appProject);
        if (appConfig.projectType !== 'application') {
            throw new Error(`Expected ${options.appProject} to be an application but got ${appConfig.projectType}`);
        }
        appProjectSourcePath = appConfig.sourceRoot;
        appMainFilePath = path.join(appProjectSourcePath, options.js ? 'main.js' : 'main.tsx');
        if (!host.exists(appMainFilePath)) {
            throw new Error(`Could not find ${appMainFilePath} during store configuration`);
        }
    }
    return {
        ...options,
        ...extraNames,
        fileName,
        constantName: (0, devkit_1.names)(options.name).constantName.toUpperCase(),
        directory,
        projectType,
        projectSourcePath: sourceRoot,
        projectModulePath: modulePath,
        appProjectSourcePath,
        appMainFilePath,
    };
}
exports.default = reduxGenerator;
