// @nx/js/src/utils/assets/copy-assets-handler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CopyAssetsHandler = exports.defaultFileEventHandler = void 0;
const minimatch_1 = require("minimatch");
const pathPosix = require("node:path/posix");
const path = require("node:path");
const fse = require("fs-extra");
const ignore_1 = require("ignore");
const fg = require("fast-glob");
const devkit_1 = require("@nx/devkit");
const client_1 = require("nx/src/daemon/client/client");
const defaultFileEventHandler = (events) => {
    const dirs = new Set(events.map((event) => path.dirname(event.dest)));
    dirs.forEach((d) => fse.ensureDirSync(d));
    events.forEach((event) => {
        if (event.type === 'create' || event.type === 'update') {
            if (fse.lstatSync(event.src).isFile()) {
                fse.copyFileSync(event.src, event.dest);
            }
        }
        else if (event.type === 'delete') {
            fse.removeSync(event.dest);
        }
        else {
            devkit_1.logger.error(`Unknown file event: ${event.type}`);
        }
    });
};
exports.defaultFileEventHandler = defaultFileEventHandler;
class CopyAssetsHandler {
    constructor(opts) {
        this.rootDir = opts.rootDir;
        this.projectDir = opts.projectDir;
        this.outputDir = opts.outputDir;
        this.callback = opts.callback ?? exports.defaultFileEventHandler;
        // TODO(jack): Should handle nested .gitignore files
        this.ignore = (0, ignore_1.default)();
        const gitignore = pathPosix.join(opts.rootDir, '.gitignore');
        const nxignore = pathPosix.join(opts.rootDir, '.nxignore');
        if (fse.existsSync(gitignore))
            this.ignore.add(fse.readFileSync(gitignore).toString());
        if (fse.existsSync(nxignore))
            this.ignore.add(fse.readFileSync(nxignore).toString());
        this.assetGlobs = opts.assets.map((f) => {
            let isGlob = false;
            let pattern;
            // Input and output directories are normalized to be relative to root
            let input;
            let output;
            let ignore = null;
            if (typeof f === 'string') {
                pattern = f;
                input = path.relative(opts.rootDir, opts.projectDir);
                output = path.relative(opts.rootDir, opts.outputDir);
            }
            else {
                isGlob = true;
                pattern = pathPosix.join(f.input, f.glob);
                input = f.input;
                output = pathPosix.join(path.relative(opts.rootDir, opts.outputDir), f.output);
                if (f.ignore)
                    ignore = f.ignore.map((ig) => pathPosix.join(f.input, ig));
            }
            return {
                isGlob,
                input,
                pattern,
                ignore,
                output,
            };
        });
    }
    async processAllAssetsOnce() {
        await Promise.all(this.assetGlobs.map(async (ag) => {
            const pattern = this.normalizeAssetPattern(ag);
            // fast-glob only supports Unix paths
            const files = await fg(pattern.replace(/\\/g, '/'), {
                cwd: this.rootDir,
                dot: true, // enable hidden files
            });
            this.callback(this.filesToEvent(files, ag));
        }));
    }
    processAllAssetsOnceSync() {
        this.assetGlobs.forEach((ag) => {
            const pattern = this.normalizeAssetPattern(ag);
            // fast-glob only supports Unix paths
            const files = fg.sync(pattern.replace(/\\/g, '/'), {
                cwd: this.rootDir,
                dot: true, // enable hidden files
            });
            this.callback(this.filesToEvent(files, ag));
        });
    }
    async watchAndProcessOnAssetChange() {
        const unregisterFileWatcher = await client_1.daemonClient.registerFileWatcher({
            watchProjects: 'all',
            includeGlobalWorkspaceFiles: true,
        }, (err, data) => {
            if (err === 'closed') {
                devkit_1.logger.error(`Watch error: Daemon closed the connection`);
                process.exit(1);
            }
            else if (err) {
                devkit_1.logger.error(`Watch error: ${err?.message ?? 'Unknown'}`);
            }
            else {
                this.processWatchEvents(data.changedFiles);
            }
        });
        return () => unregisterFileWatcher();
    }
    async processWatchEvents(events) {
        const fileEvents = [];
        for (const event of events) {
            const pathFromRoot = path.relative(this.rootDir, event.path);
            for (const ag of this.assetGlobs) {
                if ((0, minimatch_1.minimatch)(pathFromRoot, ag.pattern) &&
                    !ag.ignore?.some((ig) => (0, minimatch_1.minimatch)(pathFromRoot, ig)) &&
                    !this.ignore.ignores(pathFromRoot)) {
                    const relPath = path.relative(ag.input, pathFromRoot);
                    const destPath = relPath.startsWith('..') ? pathFromRoot : relPath;
                    fileEvents.push({
                        type: event.type,
                        src: path.join(this.rootDir, pathFromRoot),
                        dest: path.join(this.rootDir, ag.output, destPath),
                    });
                    // Match first entry and skip the rest for this file.
                    break;
                }
            }
        }
        if (fileEvents.length > 0)
            this.callback(fileEvents);
    }
    filesToEvent(files, assetGlob) {
        return files.reduce((acc, src) => {
            if (!assetGlob.ignore?.some((ig) => (0, minimatch_1.minimatch)(src, ig)) &&
                !this.ignore.ignores(src)) {
                const relPath = path.relative(assetGlob.input, src);
                const dest = relPath.startsWith('..') ? src : relPath;
                acc.push({
                    type: 'create',
                    src: path.join(this.rootDir, src),
                    dest: path.join(this.rootDir, assetGlob.output, dest),
                });
            }
            return acc;
        }, []);
    }
    normalizeAssetPattern(assetEntry) {
        return typeof assetEntry === 'string' ? assetEntry : assetEntry.pattern;
    }
}
exports.CopyAssetsHandler = CopyAssetsHandler;
