// @nx/react/src/generators/library/lib/add-linting.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addLinting = addLinting;
const eslint_1 = require("@nx/eslint");
const path_1 = require("nx/src/utils/path");
const devkit_1 = require("@nx/devkit");
const lint_1 = require("../../../utils/lint");
const eslint_file_1 = require("@nx/eslint/src/generators/utils/eslint-file");
async function addLinting(host, options) {
    if (options.linter === eslint_1.Linter.EsLint) {
        const lintTask = await (0, eslint_1.lintProjectGenerator)(host, {
            linter: options.linter,
            project: options.name,
            tsConfigPaths: [
                (0, path_1.joinPathFragments)(options.projectRoot, 'tsconfig.lib.json'),
            ],
            unitTestRunner: options.unitTestRunner,
            skipFormat: true,
            skipPackageJson: options.skipPackageJson,
            setParserOptionsProject: options.setParserOptionsProject,
            addPlugin: options.addPlugin,
        });
        if ((0, eslint_file_1.isEslintConfigSupported)(host)) {
            (0, eslint_file_1.addExtendsToLintConfig)(host, options.projectRoot, 'plugin:@nx/react');
        }
        let installTask = () => { };
        if (!options.skipPackageJson) {
            installTask = (0, devkit_1.addDependenciesToPackageJson)(host, lint_1.extraEslintDependencies.dependencies, lint_1.extraEslintDependencies.devDependencies);
        }
        return (0, devkit_1.runTasksInSerial)(lintTask, installTask);
    }
    else {
        return () => { };
    }
}
