// @nx/react/src/utils/versions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.swcPluginStyledComponentsVersion = exports.swcPluginEmotionVersion = exports.swcPluginStyledJsxVersion = exports.svgrRollupVersion = exports.rollupPluginUrlVersion = exports.sassVersion = exports.lessVersion = exports.moduleFederationEnhancedVersion = exports.moduleFederationNodeVersion = exports.typesCorsVersion = exports.corsVersion = exports.isbotVersion = exports.typesExpressVersion = exports.expressVersion = exports.autoprefixerVersion = exports.tailwindcssVersion = exports.postcssVersion = exports.tsLibVersion = exports.babelPluginStyledComponentsVersion = exports.eslintPluginReactHooksVersion = exports.eslintPluginReactVersion = exports.eslintPluginJsxA11yVersion = exports.eslintPluginImportVersion = exports.reactTestRendererVersion = exports.reactReduxVersion = exports.reduxjsToolkitVersion = exports.testingLibraryReactVersion = exports.reactRouterDomVersion = exports.styledJsxVersion = exports.emotionBabelPlugin = exports.emotionReactVersion = exports.emotionStyledVersion = exports.typesStyledComponentsVersion = exports.styledComponentsVersion = exports.babelCoreVersion = exports.babelPresetReactVersion = exports.typesNodeVersion = exports.typesReactIsVersion = exports.typesReactDomVersion = exports.typesReactVersion = exports.babelLoaderVersion = exports.swcLoaderVersion = exports.reactIsVersion = exports.reactDomVersion = exports.reactVersion = exports.nxRspackVersion = exports.nxVersion = void 0;
exports.nxVersion = require('../../package.json').version;
// Always pull the latest version until we merge rspack plugin into the repo.
exports.nxRspackVersion = '*';
exports.reactVersion = '18.3.1';
exports.reactDomVersion = '18.3.1';
exports.reactIsVersion = '18.3.1';
exports.swcLoaderVersion = '0.1.15';
exports.babelLoaderVersion = '^9.1.2';
exports.typesReactVersion = '18.3.1';
exports.typesReactDomVersion = '18.3.0';
exports.typesReactIsVersion = '18.3.0';
exports.typesNodeVersion = '18.16.9';
exports.babelPresetReactVersion = '^7.14.5';
exports.babelCoreVersion = '^7.14.5';
exports.styledComponentsVersion = '5.3.6';
exports.typesStyledComponentsVersion = '5.1.26';
exports.emotionStyledVersion = '11.11.0';
exports.emotionReactVersion = '11.11.1';
exports.emotionBabelPlugin = '11.11.0';
// WARNING: This needs to be in sync with Next.js' dependency or else there might be issues.
exports.styledJsxVersion = '5.1.2';
exports.reactRouterDomVersion = '6.11.2';
exports.testingLibraryReactVersion = '15.0.6';
exports.reduxjsToolkitVersion = '1.9.3';
exports.reactReduxVersion = '8.0.5';
exports.reactTestRendererVersion = '18.2.0';
exports.eslintPluginImportVersion = '2.27.5';
exports.eslintPluginJsxA11yVersion = '6.7.1';
exports.eslintPluginReactVersion = '7.32.2';
exports.eslintPluginReactHooksVersion = '4.6.0';
exports.babelPluginStyledComponentsVersion = '1.10.7';
exports.tsLibVersion = '^2.3.0';
exports.postcssVersion = '8.4.38';
exports.tailwindcssVersion = '3.4.3';
exports.autoprefixerVersion = '10.4.13';
// SSR and Module Federation
exports.expressVersion = '~4.18.2';
exports.typesExpressVersion = '4.17.17';
exports.isbotVersion = '^3.6.5';
exports.corsVersion = '~2.8.5';
exports.typesCorsVersion = '~2.8.12';
exports.moduleFederationNodeVersion = '~2.4.0';
exports.moduleFederationEnhancedVersion = '~0.2.3';
// style preprocessors
exports.lessVersion = '3.12.2';
exports.sassVersion = '^1.55.0';
// rollup plugins (if needed)
exports.rollupPluginUrlVersion = '^8.0.2';
exports.svgrRollupVersion = '^8.1.0';
exports.swcPluginStyledJsxVersion = '^1.5.67';
exports.swcPluginEmotionVersion = '^2.5.67';
exports.swcPluginStyledComponentsVersion = '^1.5.67';
