// @nx/react/src/generators/application/lib/add-e2e.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addE2e = addE2e;
const devkit_1 = require("@nx/devkit");
const web_1 = require("@nx/web");
const versions_1 = require("../../../utils/versions");
const has_webpack_plugin_1 = require("../../../utils/has-webpack-plugin");
const has_vite_plugin_1 = require("../../../utils/has-vite-plugin");
const find_plugin_for_config_file_1 = require("@nx/devkit/src/utils/find-plugin-for-config-file");
const target_defaults_utils_1 = require("@nx/devkit/src/generators/target-defaults-utils");
async function addE2e(tree, options) {
    const hasNxBuildPlugin = (options.bundler === 'webpack' && (0, has_webpack_plugin_1.hasWebpackPlugin)(tree)) ||
        (options.bundler === 'vite' && (0, has_vite_plugin_1.hasVitePlugin)(tree));
    if (!hasNxBuildPlugin) {
        await (0, web_1.webStaticServeGenerator)(tree, {
            buildTarget: `${options.projectName}:build`,
            targetName: 'serve-static',
            spa: true,
        });
    }
    switch (options.e2eTestRunner) {
        case 'cypress': {
            const { configurationGenerator } = (0, devkit_1.ensurePackage)('@nx/cypress', versions_1.nxVersion);
            (0, devkit_1.addProjectConfiguration)(tree, options.e2eProjectName, {
                projectType: 'application',
                root: options.e2eProjectRoot,
                sourceRoot: (0, devkit_1.joinPathFragments)(options.e2eProjectRoot, 'src'),
                targets: {},
                implicitDependencies: [options.projectName],
                tags: [],
            });
            const e2eTask = await configurationGenerator(tree, {
                ...options,
                project: options.e2eProjectName,
                directory: 'src',
                // the name and root are already normalized, instruct the generator to use them as is
                bundler: options.bundler === 'rspack' ? 'webpack' : options.bundler,
                skipFormat: true,
                devServerTarget: `${options.projectName}:${options.e2eWebServerTarget}`,
                baseUrl: options.e2eWebServerAddress,
                jsx: true,
                rootProject: options.rootProject,
                webServerCommands: hasNxBuildPlugin
                    ? {
                        default: `nx run ${options.projectName}:${options.e2eWebServerTarget}`,
                        production: `nx run ${options.projectName}:preview`,
                    }
                    : undefined,
                ciWebServerCommand: hasNxBuildPlugin
                    ? `nx run ${options.projectName}:${options.e2eCiWebServerTarget}`
                    : undefined,
                ciBaseUrl: options.bundler === 'vite' ? options.e2eCiBaseUrl : undefined,
            });
            if (options.addPlugin ||
                (0, devkit_1.readNxJson)(tree).plugins?.find((p) => typeof p === 'string'
                    ? p === '@nx/cypress/plugin'
                    : p.plugin === '@nx/cypress/plugin')) {
                let buildTarget = '^build';
                if (hasNxBuildPlugin) {
                    const configFile = options.bundler === 'webpack'
                        ? 'webpack.config.js'
                        : options.bundler === 'vite'
                            ? `vite.config.${options.js ? 'js' : 'ts'}`
                            : 'webpack.config.js';
                    const matchingPlugin = await (0, find_plugin_for_config_file_1.findPluginForConfigFile)(tree, `@nx/${options.bundler}/plugin`, (0, devkit_1.joinPathFragments)(options.appProjectRoot, configFile));
                    if (matchingPlugin && typeof matchingPlugin !== 'string') {
                        buildTarget = `^${matchingPlugin.options?.buildTargetName ?? 'build'}`;
                    }
                }
                await (0, target_defaults_utils_1.addE2eCiTargetDefaults)(tree, '@nx/cypress/plugin', buildTarget, (0, devkit_1.joinPathFragments)(options.e2eProjectRoot, `cypress.config.${options.js ? 'js' : 'ts'}`));
            }
            return e2eTask;
        }
        case 'playwright': {
            const { configurationGenerator } = (0, devkit_1.ensurePackage)('@nx/playwright', versions_1.nxVersion);
            (0, devkit_1.addProjectConfiguration)(tree, options.e2eProjectName, {
                projectType: 'application',
                root: options.e2eProjectRoot,
                sourceRoot: (0, devkit_1.joinPathFragments)(options.e2eProjectRoot, 'src'),
                targets: {},
                implicitDependencies: [options.projectName],
            });
            const e2eTask = await configurationGenerator(tree, {
                project: options.e2eProjectName,
                skipFormat: true,
                skipPackageJson: options.skipPackageJson,
                directory: 'src',
                js: false,
                linter: options.linter,
                setParserOptionsProject: options.setParserOptionsProject,
                webServerCommand: `${(0, devkit_1.getPackageManagerCommand)().exec} nx run ${options.projectName}:${options.e2eCiWebServerTarget}`,
                webServerAddress: options.e2eCiBaseUrl,
                rootProject: options.rootProject,
                addPlugin: options.addPlugin,
            });
            if (options.addPlugin ||
                (0, devkit_1.readNxJson)(tree).plugins?.find((p) => typeof p === 'string'
                    ? p === '@nx/playwright/plugin'
                    : p.plugin === '@nx/playwright/plugin')) {
                let buildTarget = '^build';
                if (hasNxBuildPlugin) {
                    const configFile = options.bundler === 'webpack'
                        ? 'webpack.config.js'
                        : options.bundler === 'vite'
                            ? `vite.config.${options.js ? 'js' : 'ts'}`
                            : 'webpack.config.js';
                    const matchingPlugin = await (0, find_plugin_for_config_file_1.findPluginForConfigFile)(tree, `@nx/${options.bundler}/plugin`, (0, devkit_1.joinPathFragments)(options.appProjectRoot, configFile));
                    if (matchingPlugin && typeof matchingPlugin !== 'string') {
                        buildTarget = `^${matchingPlugin.options?.buildTargetName ?? 'build'}`;
                    }
                }
                await (0, target_defaults_utils_1.addE2eCiTargetDefaults)(tree, '@nx/playwright/plugin', buildTarget, (0, devkit_1.joinPathFragments)(options.e2eProjectRoot, `playwright.config.ts`));
            }
            return e2eTask;
        }
        case 'none':
        default:
            return () => { };
    }
}
