// @nx/nest/src/generators/application/application.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.applicationGenerator = applicationGenerator;
exports.applicationGeneratorInternal = applicationGeneratorInternal;
const devkit_1 = require("@nx/devkit");
const node_1 = require("@nx/node");
const init_1 = require("../init/init");
const lib_1 = require("./lib");
const ensure_dependencies_1 = require("../../utils/ensure-dependencies");
async function applicationGenerator(tree, rawOptions) {
    return await applicationGeneratorInternal(tree, {
        addPlugin: false,
        projectNameAndRootFormat: 'derived',
        ...rawOptions,
    });
}
async function applicationGeneratorInternal(tree, rawOptions) {
    const options = await (0, lib_1.normalizeOptions)(tree, rawOptions);
    const tasks = [];
    const initTask = await (0, init_1.initGenerator)(tree, {
        skipPackageJson: options.skipPackageJson,
        skipFormat: true,
    });
    tasks.push(initTask);
    const nodeApplicationTask = await (0, node_1.applicationGenerator)(tree, (0, lib_1.toNodeApplicationGeneratorOptions)(options));
    tasks.push(nodeApplicationTask);
    (0, lib_1.createFiles)(tree, options);
    (0, lib_1.updateTsConfig)(tree, options);
    if (!options.skipPackageJson) {
        tasks.push((0, ensure_dependencies_1.ensureDependencies)(tree));
    }
    if (!options.skipFormat) {
        await (0, devkit_1.formatFiles)(tree);
    }
    return (0, devkit_1.runTasksInSerial)(...tasks);
}
exports.default = applicationGenerator;
