// @nx/nest/src/generators/interceptor/interceptor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.interceptorGenerator = interceptorGenerator;
exports.interceptorGeneratorInternal = interceptorGeneratorInternal;
const utils_1 = require("../utils");
async function interceptorGenerator(tree, rawOptions) {
    await interceptorGeneratorInternal(tree, {
        nameAndDirectoryFormat: 'derived',
        ...rawOptions,
    });
}
async function interceptorGeneratorInternal(tree, rawOptions) {
    const options = await normalizeInterceptorOptions(tree, rawOptions);
    return (0, utils_1.runNestSchematic)(tree, 'interceptor', options);
}
exports.default = interceptorGenerator;
async function normalizeInterceptorOptions(tree, options) {
    const normalizedOptions = await (0, utils_1.normalizeOptions)(tree, 'interceptor', '@nx/nest:interceptor', options);
    return {
        ...normalizedOptions,
        language: options.language,
        spec: (0, utils_1.unitTestRunnerToSpec)(options.unitTestRunner),
    };
}
