// @nx/react/src/rules/update-module-federation-project.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateModuleFederationProject = updateModuleFederationProject;
const devkit_1 = require("@nx/devkit");
const versions_1 = require("../utils/versions");
const maybe_js_1 = require("../utils/maybe-js");
function updateModuleFederationProject(host, options) {
    const projectConfig = (0, devkit_1.readProjectConfiguration)(host, options.projectName);
    projectConfig.targets.build.options = {
        ...projectConfig.targets.build.options,
        main: (0, maybe_js_1.maybeJs)(options, `${options.appProjectRoot}/src/main.ts`),
        webpackConfig: `${options.appProjectRoot}/webpack.config.${options.typescriptConfiguration && !options.js ? 'ts' : 'js'}`,
    };
    projectConfig.targets.build.configurations.production = {
        ...projectConfig.targets.build.configurations.production,
        webpackConfig: `${options.appProjectRoot}/webpack.config.prod.${options.typescriptConfiguration && !options.js ? 'ts' : 'js'}`,
    };
    // If host should be configured to use dynamic federation
    if (options.dynamic) {
        const pathToProdWebpackConfig = (0, devkit_1.joinPathFragments)(projectConfig.root, `webpack.prod.config.${options.typescriptConfiguration && !options.js ? 'ts' : 'js'}`);
        if (host.exists(pathToProdWebpackConfig)) {
            host.delete(pathToProdWebpackConfig);
        }
        delete projectConfig.targets.build.configurations.production?.webpackConfig;
    }
    projectConfig.targets.serve.executor =
        '@nx/react:module-federation-dev-server';
    projectConfig.targets.serve.options.port = options.devServerPort;
    // `serve-static` for remotes that don't need to be in development mode
    projectConfig.targets['serve-static'] = {
        executor: '@nx/web:file-server',
        defaultConfiguration: 'production',
        options: {
            buildTarget: `${options.projectName}:build`,
            watch: false,
            port: options.devServerPort,
        },
        configurations: {
            development: {
                buildTarget: `${options.projectName}:build:development`,
            },
            production: {
                buildTarget: `${options.projectName}:build:production`,
            },
        },
    };
    (0, devkit_1.updateProjectConfiguration)(host, options.projectName, projectConfig);
    return (0, devkit_1.addDependenciesToPackageJson)(host, {}, { '@nx/web': versions_1.nxVersion });
}
