// @nx/react/src/generators/host/lib/add-module-federation-files.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addModuleFederationFiles = addModuleFederationFiles;
const devkit_1 = require("@nx/devkit");
const maybe_js_1 = require("../../../utils/maybe-js");
function addModuleFederationFiles(host, options, defaultRemoteManifest) {
    const templateVariables = {
        ...(0, devkit_1.names)(options.name),
        ...options,
        static: !options?.dynamic,
        tmpl: '',
        remotes: defaultRemoteManifest.map(({ name, port }) => {
            return {
                ...(0, devkit_1.names)(name),
                port,
            };
        }),
    };
    const projectConfig = (0, devkit_1.readProjectConfiguration)(host, options.name);
    const pathToMFManifest = (0, devkit_1.joinPathFragments)(projectConfig.sourceRoot, 'assets/module-federation.manifest.json');
    // Module federation requires bootstrap code to be dynamically imported.
    // Renaming original entry file so we can use `import(./bootstrap)` in
    // new entry file.
    host.rename((0, devkit_1.joinPathFragments)(options.appProjectRoot, (0, maybe_js_1.maybeJs)(options, 'src/main.tsx')), (0, devkit_1.joinPathFragments)(options.appProjectRoot, (0, maybe_js_1.maybeJs)(options, 'src/bootstrap.tsx')));
    (0, devkit_1.generateFiles)(host, (0, devkit_1.joinPathFragments)(__dirname, `../files/${options.js ? 'common' : 'common-ts'}`), options.appProjectRoot, templateVariables);
    const pathToModuleFederationFiles = options.typescriptConfiguration
        ? 'module-federation-ts'
        : 'module-federation';
    // New entry file is created here.
    (0, devkit_1.generateFiles)(host, (0, devkit_1.joinPathFragments)(__dirname, `../files/${pathToModuleFederationFiles}`), options.appProjectRoot, templateVariables);
    function deleteFileIfExists(host, filePath) {
        if (host.exists(filePath)) {
            host.delete(filePath);
        }
    }
    function processWebpackConfig(options, host, fileName) {
        const pathToWebpackConfig = (0, devkit_1.joinPathFragments)(options.appProjectRoot, fileName);
        deleteFileIfExists(host, pathToWebpackConfig);
    }
    if (options.typescriptConfiguration) {
        processWebpackConfig(options, host, 'webpack.config.js');
        processWebpackConfig(options, host, 'webpack.config.prod.js');
    }
    if (options.dynamic) {
        processWebpackConfig(options, host, 'webpack.config.prod.js');
        processWebpackConfig(options, host, 'webpack.config.prod.ts');
        if (!host.exists(pathToMFManifest)) {
            host.write(pathToMFManifest, `{
        ${defaultRemoteManifest
                .map(({ name, port }) => `"${name}": "http://localhost:${port}"`)
                .join(',\n')}
          }`);
        }
    }
}
